src/NBitcoin/ConsensusFactory.cs

```diff
@@ -0,0 +1,239 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Reflection;
+
+namespace NBitcoin
+{
+    public class ConsensusFactory
+    {
+        private readonly ConcurrentDictionary<Type, bool> isAssignableFromBlockHeader = new ConcurrentDictionary<Type, bool>();
+        private readonly TypeInfo blockHeaderType = typeof(BlockHeader).GetTypeInfo();
+
+        private readonly ConcurrentDictionary<Type, bool> isAssignableFromBlock = new ConcurrentDictionary<Type, bool>();
+        private readonly TypeInfo blockType = typeof(Block).GetTypeInfo();
+
+        private readonly ConcurrentDictionary<Type, bool> isAssignableFromTransaction = new ConcurrentDictionary<Type, bool>();
+        private readonly TypeInfo transactionType = typeof(Transaction).GetTypeInfo();
+
+        protected bool IsBlockHeader<T>()
+        {
+            return this.IsAssignable<T>(this.blockHeaderType, this.isAssignableFromBlockHeader);
+        }
+
+        protected bool IsBlock<T>()
+        {
+            return this.IsAssignable<T>(this.blockType, this.isAssignableFromBlock);
+        }
+
+        protected bool IsTransaction<T>()
+        {
+            return this.IsAssignable<T>(this.transactionType, this.isAssignableFromTransaction);
+        }
+
+        /// <summary>
+        /// Represents the parent of this class.
+        /// For simplicity (and migration from NBitcoin) I made this a property however this design can be revised later.
+        /// </summary>
+        public Consensus Consensus { get; set; }
+
+        private bool IsAssignable<T>(TypeInfo type, ConcurrentDictionary<Type, bool> cache)
+        {
+            bool isAssignable = false;
+            if (!cache.TryGetValue(typeof(T), out isAssignable))
```

Comment:
can do `out bool isAssignable` instead of declaring it before.

Reply:
yes done